### ISSUE TITLE
`contentFor` hooks not called for nested addons

### ISSUE THREAD

**Issue by alexlafroscia** [ember-cli/ember-cli#8077]
I've been re-arranging some of our addons and noticed that the `contentFor` hooks for nested addons don't seem to be run.

Our app, `core-ui`, had `core-components` as a dependency. That addon has an addon in `dependencies` that has some `contentFor` hooks that are not fired.

The application finds the components within the nested addon, so _some_ features are working okay. But, for whatever reason, when [`configReplacePatterns`](https://github.com/ember-cli/ember-cli/blob/9d7fd5f3fea476944a4e9ddc38f1f7da57fee0ca/lib/utilities/ember-app-utils.js#L161) is called, the nested addon is missing from the passed-in list.

---

Output from `ember version --verbose && npm --version && yarn --version`:
```
❯ yarn ember version --verbose && npm --version && yarn --version
yarn run v1.9.4
$__PATH_TO_PROJECT__/node_modules/.bin/ember version --verbose
ember-cli: 3.3.0
http_parser: 2.7.0
node: 8.10.0
v8: 6.2.414.50
uv: 1.19.1
zlib: 1.2.11
ares: 1.10.1-DEV
modules: 57
nghttp2: 1.25.0
openssl: 1.0.2n
icu: 60.1
unicode: 10.0
cldr: 32.0
tz: 2017c
os: darwin x64
✨  Done in 16.23s.
5.6.0
1.9.4
```

**Comment by stefanpenner** [comment]
This is expected behavior.

**Comment by stefanpenner** [comment]
Yes, the nested add-on concept was sorta pushed without much design. Some hooks are omitted, as they are incoherent in a duplicate add-on world. That being said, duplicate addons themselves are somewhat incoherent. 

If we are to change this, it would be part of a largely "duplicate/nested addon" re-think.

**Comment by alexlafroscia** [comment]
What would be the appropriate short-term "fix" here? Maybe just better documentation around some of these "gotchas"? As it stands right now, from an addon author's perspective, it's hard to know whether they can include a dependent addon in their own `dependencies` or whether they should use the default blueprint to install it for the host as well.

I, personally, _really_ dislike the "install another addon as a side effect" approach that we have right now in many cases. It makes cleanup when removing the original addon difficult, because they have introduced a number of dependencies that would also need to be explicitly removed from the host application.

**Comment by mydea** [comment]
We had the same issue in one of our addons, and ended up using this pattern from ember-power-select/ember-basic-dropdown:

https://github.com/cibernox/ember-power-select/blob/master/index.js#L37

Note that you'll need to guard against calling the contentFor hook multiple times (https://github.com/cibernox/ember-basic-dropdown/blob/master/index.js#L27).

Not necessarily ideal, but it does work for now.

**Comment by alexlafroscia** [comment]
Thanks for that context, @mydea! That's an interesting idea, for cases where you know which addon you need to invoke the `contentFor` hooks for.

**Comment by kellyselden** [comment]
I made https://github.com/CrowdStrike/ember-cli-content-for-nested  as a workaround for now. Use it if you wish.

**Comment by stefanpenner** [comment]
@kellyselden that is dangerous, I avoid advise against.

**Comment by RuslanZavacky** [comment]
@stefanpenner can you please elaborate - what is dangerous? Concrete case "this will happen if you do so" would be very much appreciated. We struggle with addon nested issue for the 3+ years so far :) Haven't seen any better workaround for it..

**Comment by stefanpenner** [comment]
@ruslanzavacky nested add-ons result in duplicates of differing versions, which result in collisions in places like this. These collisions are not resolvable by the host application, and can occur within a Semver safe bump. The hazard here is very high.


Although we already have a collision problem introduced by addons code, adding more such opportunities will only worsen the pain.

also although  addon tree collisions (for JS modules) occurs today it is actually solvable via import/export rewriting (planned work), but the collisions caused by content-for (and similar global constructs) do not have obvious conflict resolution strategies, at-least not without more in-depth design.

This design was skipped entirely when nested addons were added, resulting in the current pain you currently face. 

To avoid expanding the problem further, design must be a pre-req to any further loosening of the constraints here.

**Comment by NullVoxPopuli** [comment]
I just ran in to this, I would like `contentFor` to be used in nested addons :-\

what's a situation in which a duplicate occurs?

_surely_ there is a way to only run a `contentFor` hook once per addon?

**Comment by NullVoxPopuli** [comment]
my use case is that:

I have an app
 - which has an addon
    - which depends on another addon which has contentFor <-- yet this contentFor does not run
